protocols/autonatv2/src/client/behaviour.rs

```diff
@@ -0,0 +1,294 @@
+                dial_request: DialRequest { nonce, addrs },
+                suspicious_addr,
+                reachable_addr,
+            }))) => {
+                if self.pending_nonces.remove(&nonce) {
+                    tracing::debug!(
+                        "server reported reachbility, but didn't actually reached this node."
+                    );
+                    return;
+                }
+                if !suspicious_addr.is_empty() {
+                    tracing::trace!(
+                        "server reported suspicious addresses: {:?}",
+                        suspicious_addr
+                    );
+                }
+                self.pending_events.extend(
+                    addrs
+                        .into_iter()
+                        .take_while(|addr| addr != &reachable_addr)
+                        .map(ToSwarm::ExternalAddrExpired),
+                );
+                self.pending_events
+                    .push_back(ToSwarm::ExternalAddrConfirmed(reachable_addr));
+            }
+            Either::Left(dial_request::ToBehaviour::TestCompleted(
+                Err(dial_request::Error::UnableToConnectOnSelectedAddress { addr: Some(addr) })
+            ))
+            | Either::Left(dial_request::ToBehaviour::TestCompleted(
+                Err(dial_request::Error::FailureDuringDialBack { addr: Some(addr) })
+            )) => {
+                self.pending_events
+                    .push_back(ToSwarm::ExternalAddrExpired(addr));
+            }
+            Either::Left(dial_request::ToBehaviour::TestCompleted(Err(err))) => {
+                tracing::debug!("Test failed: {:?}", err);
+            }
+        }
+    }
+
+    fn poll(
+        &mut self,
+        _cx: &mut Context<'_>,
+    ) -> Poll<ToSwarm<Self::ToSwarm, <Handler as ConnectionHandler>::FromBehaviour>> {
+        let pending_event = self.poll_pending_events();
+        if pending_event.is_ready() {
+            return pending_event;
+        }
+        if self.ticker.ready() && !self.known_servers.is_empty() {
+            let mut entries = self.address_candidates.drain().collect::<Vec<_>>();
+            entries.sort_unstable_by_key(|(_, count)| *count);
+            let addrs = entries
+                .into_iter()
+                .rev()
+                .map(|(addr, _)| addr)
+                .take(self.config.max_addrs_count)
+                .collect::<Vec<_>>();
+            let peers = if self.known_servers.len() < self.config.test_server_count {
+                self.known_servers.clone()
+            } else {
+                self.known_servers
+                    .choose_multiple(&mut self.rng, self.config.test_server_count)
+                    .copied()
+                    .collect()
+            };
+            for peer in peers {
+                let nonce = self.rng.gen();
+                let req = DialRequest {
+                    nonce,
+                    addrs: addrs.clone(),
```

Comment:
The server respects the order of this list, right? With the current logic, we will likely test the same address with many different servers. Whilst that gives us confidence, it means we are much less likely to test the other addresses.

What do you think of changing this to only send a single address to each server and picking a server randomly for each address that we want to test? Or perhaps we should send the same request to two randomly selected autonat servers?

Reply:
Honestly good call. I will go with the former.